### PR TITLE
FXML-635: GlobalAveragePool2D op

### PR DESCRIPTION
**Pull request by BraunPhilipp**
* Add `GlobalAveragePool2D` op to `LinAlg`
* Update LLVM formatting (sorry; hope my formatter picked that up correctly)
* Update unfusing to keep correct semantics

### CODE REVIEW

**Comment by BraunPhilipp** [comment]
Just noticed that `.clang-format` broke for me again. Will fix that in the office.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 1]
```diff
@@ -680,21 +680,34 @@ LogicalResult SoftmaxOp::verify() {
...
+//===----------------------------------------------------------------------===//
+// GlobalAveragePool2D
+//===----------------------------------------------------------------------===//
+
+LogicalResult GlobalAveragePool2DOp::verify() {
+  auto inputTensor = input().getType().dyn_cast<RankedTensorType>();
+  auto outputTensor = result().getType().dyn_cast<RankedTensorType>();
+  if (inputTensor.getRank() != outputTensor.getRank())
+    return emitOpError("rank in input and output tensor needs to match");
+  return success();
```
Comment (KFAFSP) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:692]: Are we not going to check the rest of the shape? That's asking for trouble once we materialize any tiling.

**Comment by KFAFSP** [comment]
Admittedly, that stuff hadn't been formatted before, but this makes the or hard to check.

File: mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp [hunk 2]
```diff
@@ -357,281 +303,293 @@ Value unfuseActivation(Conv2DReluMaxpoolOp op, Value ifm,
...
-    iteratorTypes.back() = "reduction";
-
-    return builder.create<linalg::GenericOp>(
-        loc,
-        /*resultTensorTypes=*/sumTy,
-        /*inputs=*/in, 
-        /*outputs=*/accu, 
-        indexingMaps, 
-        iteratorTypes, 
+    rewriter.replaceOpWithNewOp<linalg::GenericOp>(
```
[the remark this answers is not in the document]
Comment (KFAFSP) [reply mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp:558]: Also, you want `divf` instead. And actually, you need to make this work for integer types too, so this is wrong still.

File: mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp [hunk 3]
```diff
@@ -357,281 +303,293 @@ Value unfuseActivation(Conv2DReluMaxpoolOp op, Value ifm,
...
+    Value accu =
+        rewriter.create<tensor::SplatOp>(loc, zero, resultTy).getResult();
+    Value sum = rewriter
+                    .create<linalg::PoolingNchwSumOp>(
+                        loc, ArrayRef<Type>{resultTy},
+                        ValueRange{input, kernel}, accu, attributes)
+                    .getResult(0);
+    Value numElements = rewriter.create<arith::ConstantOp>(
+        loc, resultTy.getElementType(),
+        rewriter.getFloatAttr(resultTy.getElementType(),
```
Comment (KFAFSP) [inline mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp:539]: Does not work for integer element types.

**Review by ljfitz — APPROVED** [review]

File: mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (BraunPhilipp) [reply mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp:558]: This probably got lost in the formatting changes. Created a separate formatting PR.

File: mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp (hunk 3 above)
Comment (BraunPhilipp) [reply mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp:539]: I have noticed the same for other unfusing passes. Will appropriately `assert` here.

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgOps.td [hunk 4]
```diff
@@ -198,6 +198,17 @@ def Linalg_SoftmaxOp : Linalg_Op<"softmax", [NoSideEffect]> {
   let hasVerifier = 1;
 }
 
+def Linalg_GlobalAveragePool2DOp : Linalg_Op<"globalaveragepool2d", [NoSideEffect]> {
+  let arguments = (ins AnyTensor:$input);
+  let results = (outs AnyTensor:$result);
+  let summary = "linalg globalaveragepool2d operation";
+  let description = [{
+    This implements the globalaveragepool2d operator.
```
[the remark this answers is not in the document]
Comment (BraunPhilipp) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgOps.td:206]: Added some more information, but noticed the same for softmax.

**Comment by BraunPhilipp** [comment]
Verified using `unfuse.mlir` and custom linalg execution test.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp (hunk 1 above)
Comment (BraunPhilipp) [reply mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:697]: done!

**Pull request merged**